Add SwapAndBridge feature

ricott1 opened the pull request:
Closes #191.

Implement swap-and-bridge feature with the `SwapAndBridge` contract. This contract exposes the `swapAndBridgeToWithMinimumAmount` public function by which users can swap ETH for a target LST and bridge it directly to an L2 while also specifying the minimum amount of LST to receive. The contract also exposes some convenience functions and a receive fallback allowing the interaction to happen just by sending ETH to the contract.

A new `SwapAndBridge` contract should be deployed for each LST protocol and target L2. The contract will be created with the L1 LST token, the L1 bridge, and the target L2 token as parameters. In this PR we add integration tests for Lido and Diva LST protocols as examples.

ricott1 committed:
Add SwapAndBridge contract and Lido deployment and tests

ricott1 committed:
SwapAndBridge: Remove duplicate variables, add check against 0 address in constructor, and add slippage check

ricott1 committed:
SwapAndBridge: Remove duplicate variables, add check against 0 address in constructor, and add slippage check

ricott1 committed:
Added e2e tests.

ricott1 committed:
clean up deploy.sh

ricott1 committed:
SwapAndBridge.sol: add docs to constructor

ricott1 committed:
Clean up test

ricott1 committed:
update .env.example

ricott1 committed:
Clean up deployment script

ricott1 committed:
Clean up env file

ricott1 committed:
Add more balance for tests

ricott1 committed:
Add unit tests

ricott1 committed:
forge install: solady

v0.0.180

ricott1 committed:
remove unused contracts, update tests

ricott1 committed:
automate deployment

ricott1 committed:
fix env.example

ricott1 committed:
cleanup SwapAndBridge tests

ricott1 committed:
address review

ricott1 committed:
update contract license

matjazv reviewed (changes requested):
Nice work!
Just some minor suggestions.

ricott1 committed:
address review

matjazv reviewed (changes requested):
Looks good!
Just few small comments left.

ricott1 committed:
Fix E2E tests for Diva

ricott1 committed:
Minor fixes to Utils.sol

ricott1 committed:
Update Utils.sol with SwapAndBridge functionalities

ricott1 committed:
Fix import in Utils.sol

ricott1 committed:
Fix import in Utils.sol

ricott1 committed:
Fix variable usage

ricott1 committed:
Move env variables to contract constants to fix CI

ricott1 committed:
Fix formatting

ricott1 committed:
Remove use of env variables

ricott1 committed:
remove unused TOKEN_HOLDER_PRIV_KEY variable

ricott1 committed:
Save temporary data files to local storage

ricott1 committed:
fix fmt

ricott1 committed:
Move e2e tests to script

ricott1 committed:
Move integration tests to script; add unit tests

ricott1 committed:
remove old test folder

ricott1 committed:
Fix integration tests

ricott1 committed:
remove unnecessary function

ricott1 committed:
Add new unit tests

ricott1 committed:
Add mock contracts

ricott1 committed:
Merge branch 'main' into feature/swapAndBridge

ricott1 committed:
add tests for L2WdivETH

ricott1 committed:
Merge branch 'feature/swapAndBridge' of https://github.com/LiskHQ/lisk-contracts-private into feature/swapAndBridge

Commit:
Add AccessControl to VestingWallet

Commit:
Add NatSpec comments

Commit:
Update contract owner

Commit:
Added NatSpec and beneficiary check

ricott1 committed:
Remove redundant mappings

ricott1 committed:
Improve contract interface

ricott1 committed:
Increase test coverage

matjazv committed:
Update data for Testnet migration (#135)

Merkle root was added and unneeded account removed from accounts_2.json

Commit:
Add address checking to contract

Commit:
Merge branch 'main' of github.com:LiskHQ/lisk-contracts-private into 132-separate-beneficiary-and-owner-in-vesting

Commit:
Update error message

Commit:
Add virtual to functions

Commit:
Separate beneficiary and owner in VestingWallet (#133)

### What was the problem?

This PR resolves #132

### How was it solved?

- AccessControl introduced to Vesting Wallet, with 2-step transfer
- Terminology: `contractAdmin` can upgrade the contract, while `owner`
is the beneficiary of the Vesting. It is due to the design of
Openzeppelin contract, which by default both are `owner`. Adding new
role as `beneficiary` would cause more functions needed to be
overridden.

### How was it tested?
- Test case updated

Commit:
Vesting Wallet on Ethereum L1 Mainnet (#123)

* Added L1VestingWallet, added testnet Vesting Plans, Added L1 Vesting deployment script

* Updated vesting plans script

* Update vesting data

* Add explanation to L1VestingWallet

* Update format

* Update READMEs

* Update vesting amount

* Update vesting amounts

* Update devnet data

* Update according to suggestion

* Update L1 contract code and deployment scripts

* Update README

* Update PR review

ricott1 committed:
Merge branch 'main' into feature/swapAndBridge

ricott1 committed:
merge main

ricott1 committed:
Use OptimismMintableERC20Factory for Diva ERC20

ricott1 committed:
remove optimism library

ricott1 committed:
Merge branch 'main' into feature/swapAndBridge

matjazv committed:
Add Airdrop contract Testnet deployment data (#159)

* Set 6_deployUSDTContract.sh script file as executable

* Set 5_deployAirdropContract.sh script file as executable

* Add Airdrop contract deployment data

* Fix GitHub CI regarding Forge version

ricott1 committed:
remove unnecessary contracts

ricott1 committed:
add deployment scripts

Commit:
Added L2VotingPowerPaused and L2GovernorPaused contract with test cases (#165)

* Added L2VotingPowerPaused and L2GovernorPaused contract with test cases

* Add deployment script

* Update git submodules

* Add foundry.toml

* Change version function to be independent

* Update comments

* Update gitsubmodule, added Custom Errors

* Rearrange functions according to requirements

* pump forge-std to 1.8.2

* Add onReceive ERC721 and ERC1155 check

* Test adjustVotingPower is unpaused

* Add deployPausedDAO.sh

* Add forge clean to deploy script

* Clean before every forge script runs

* Used `../../` to navigate paused script

* rename assertInitParamsEq

ricott1 committed:
fixing e2e

ricott1 committed:
fix e2e tests

ricott1 committed:
remove optimism lib

ricott1 committed:
fix e2e tests

matjazv committed:
Merkle root for Airdrop contract is set inside deployment script (#161)

Set Merkle root on Airdrop contract deployment

ricott1 committed:
Merge branch 'main' into feature/swapAndBridge

matjazv committed:
Add automated toolings for testing contracts (#151)

* Add Slither support in GH actions

* Add slither-mutate support for local execution

* Add readme for test directory

* Fix rebasing with development

* Add Echidna support for local execution

* Fix spelling typo

ricott1 committed:
fix env files

ricott1 committed:
merge development

matjazv commented on .env.example:
Inside `.env.testnet` and `.env.mainnet` there is a variable `L2_STANDARD_BRIDGE_ADDR` present.
Should it also be inside `env.example` file?

ricott1 replied:
added (the USDT variable was also missing so I added it)

ricott1 committed:
merge

ricott1 committed:
Merge branch 'development' into feature/swapAndBridge

matjazv commented on .env.testnet:
```diff
@@ -79,3 +82,24 @@ L2_FORK_RPC_URL=https://optimism-goerli.infura.io/v3/YOUR_INFURA_KEY
+L1_LIDO_BRIDGE_ADDR=0xdDDbC273a81e6BC49c269Af55d007c08c005ea56
+
+# L2 wdivETH token contract
+L2_TOKEN_ADDR_DIVA=0x
```
Why is this address empty?
Others are filled in for `testnet`.

ricott1 replied:
this address should come from the deployed contract in script 7. Maybe a better way would be to get it from the deployments folder but I thought it would be nice to have it in the env file for the integration tests and also to be in line with what we do with USDT. Once we an an official Sepolia deployment, I can just fill it in

matjazv replied:
Sound good!

ricott1 committed:
forge install: openzeppelin-contracts

v5.0.2

ricott1 committed:
forge install: openzeppelin-contracts-upgradeable

v5.0.2

ricott1 committed:
forge install: openzeppelin-contracts

01ef448981be9d20ca85f2faf6ebdf591ce409f3

ricott1 committed:
forge install: openzeppelin-contracts-upgradeable

fbdb824a735891908d5588b28e0da5852d7ed7ba

ricott1 committed:
forge install: forge-std

v1.8.2

matjazv commented on .github/workflows/pr.yaml:
Please revert this file.

matjazv commented on script/7_deployWdivETHContract.sh:
```diff
@@ -0,0 +1,17 @@
+echo "Done."
+
+echo "Deploying OptimismMintableERC20 wdivETH token smart contract..."
+# Use flag --nonce to specify suitable nonce in case of nonce mismatch
```
Do we need this comment?

ricott1 replied:
removed

matjazv commented on script/7_deployWdivETHContract.sh:
```diff
@@ -0,0 +1,17 @@
+
+echo "Deploying OptimismMintableERC20 wdivETH token smart contract..."
+# Use flag --nonce to specify suitable nonce in case of nonce mismatch
+cast send --rpc-url $L2_RPC_URL "0x4200000000000000000000000000000000000012" "createOptimismMintableERC20WithDecimals(address,string,string,uint8)" $REMOTE_TOKEN_ADDR_WDIVETH "Wrapped Diva Ether Token" wdivETH 18 -i
```
`echo "Done."` and blank last line is missing.

matjazv replied:
Blank line is optional to insert.

ricott1 replied:
done

ricott1 committed:
cleanup

matjazv commented on script/8_deploySwapAndBridge.sh:
```diff
@@ -0,0 +1,36 @@
+echo "Done."
+
+echo "Deploying and if enabled verifying SwapAndBridge smart contract for DIVA protocol..."
+forge script --rpc-url="$L1_RPC_URL" --broadcast --verify --verifier blockscout --verifier-url $L1_VERIFIER_URL -vvvv script/swap_and_bridge/contracts/SwapAndBridge.s.sol:SwapAndBridgeScript $L1_STANDARD_BRIDGE_ADDR $L1_TOKEN_ADDR_DIVA $L2_TOKEN_ADDR_DIVA --sig 'run(address,address,address)' 
```
Why is this needed to be passed to the `forge script`?
`$L1_STANDARD_BRIDGE_ADDR $L1_TOKEN_ADDR_DIVA $L2_TOKEN_ADDR_DIVA`
Isn't this taken from `.env` file inside `SwapAndBridge.s.sol` deployment` script?

ricott1 replied:
I made the `SwapAndBridge.s.sol` agnostic to the protocol for which it is used and just accept the relevant parameter in the `run` function.

matjazv replied:
Why are in this case those variables inside `.env.*` files?

matjazv replied:
Inside `SwapAndBridgeScript` it's not checked if those variables are empty addresses or not.
It would be nice to check that.

On the other hand, all our other deployment scripts are also agnostic to which network they are deployed.
All environment variables are read and checked from `.env` file inside deployment scripts like here:
https://github.com/LiskHQ/lisk-contracts/blob/development/script/contracts/L2/L2Airdrop.s.sol#L26-#L27

In this case `forge script` command has also a smaller list of arguments.

What's your opinion about this?

ricott1 replied:
I will add the address check. Regarding the env variables, what I meant is that within the same network, you can deploy this contract for different protocols (e.g. Lido and Diva). Using the .env variables in the contract would still require to have a separate script for each protocol for which you want to deploy the contract (which in general we don't know a priori).

ricott1 replied:
the addresses are also check in the actual contract, should we still add the extra check in the script contract?

matjazv replied:
To be unified with other deployment scripts inside this repo, I would say it would be nice if also script would check if valid arguments are passed to the code.

matjazv replied:
Ok, I see.
Same deployment script is used for two different protocols (set of smart contracts).

ricott1 replied:
I think I should write a Readme in the swap and bridge folder anyway. Maybe I could explain there

matjazv commented on script/8_deploySwapAndBridge.sh:
```diff
@@ -0,0 +1,36 @@
+echo "Done."
+
+echo "Deploying and if enabled verifying SwapAndBridge smart contract for LIDO protocol..."
+forge script --rpc-url="$L1_RPC_URL" --broadcast --verify --verifier blockscout --verifier-url $L1_VERIFIER_URL -vvvv script/swap_and_bridge/contracts/SwapAndBridge.s.sol:SwapAndBridgeScript  $L1_LIDO_BRIDGE_ADDR $L1_TOKEN_ADDR_LIDO $L2_TOKEN_ADDR_LIDO --sig 'run(address,address,address)'
```
Same as above for `$L1_LIDO_BRIDGE_ADDR $L1_TOKEN_ADDR_LIDO $L2_TOKEN_ADDR_LIDO`.

ricott1 committed:
cleanup

ricott1 committed:
address Slither and Matjaz comments

ricott1 committed:
add extra deployment checks

matjazv commented on script/swap_and_bridge/contracts/SwapAndBridge.s.sol:
```diff
@@ -8,13 +8,16 @@ import { SwapAndBridge } from "src/L1/SwapAndBridge.sol";
     /// @notice This function deploys the SwapAndBridge contract.
-    function run(address bridgeAddress, address l1Token, address l2Token) public {
+    function run(address _l1Bridge, address _l1Token, address _l2Token) public {
+        require(_l1Bridge != address(0), "Invalid L1 bridge address.");
+        require(_l1Token != address(0), "Invalid L1 token address.");
+        require(_l2Token != address(0), "Invalid L2 token address.");
```
I think `assert` should be used instead of `require` inside scripts.

matjazv commented on script/swap_and_bridge/example/e2e_test.sh:
```diff
@@ -0,0 +1,23 @@
+echo "Done."
+
+echo "Navigating to the root directory of the project..."
+cd ../
```
```suggestion
cd ../../../
```

ricott1 replied:
I usually run every script from the script folder, but I agree this is unclear. If I clarify that in the README would that be fine?

matjazv replied:
All scripts inside `script` folder and its subfolders are currently executed within root `script` folder. See example:
https://github.com/LiskHQ/lisk-contracts/blob/development/script/paused/deployPausedClaim.sh#L8
Would be nice to use same pattern for scripts inside this PR.

matjazv replied:
Currently all scripts inside `script` folder and its subfolders are executed within a folder where the script is located. See example:
https://github.com/LiskHQ/lisk-contracts/blob/development/script/paused/deployPausedClaim.sh#L8
Would be nice if you use this pattern also inside this PR.

matjazv commented on script/swap_and_bridge/example/integration_test.sh:
```diff
@@ -0,0 +1,21 @@
+echo "Done."
+
+echo "Navigating to the root directory of the project..."
+cd ../
```
```suggestion
cd ../../../
```

matjazv commented on foundry.toml:
This file like `pr.yaml` may be reverted because no changes were done.

matjazv commented:
Folder `example` inside `script/swap_and_bridge/example` should be renamed to something more expressive.
Inside of it are scripts which executes different kind of test.

matjazv commented on script/swap_and_bridge/example/README.md:
```diff
@@ -0,0 +1,11 @@
+Tests in this folder are run as script as they require to be run in a Sepolia forked environment, since they depend on the contracts owned and deployed by Lido and Diva teams.
+
+Before running them, start an L1 and L2 fork node using the shell scripts `runL1TestNetwork.sh` and `runL2TestNetwork.sh` respectively.
```
```suggestion
Before running them, start an L1 and L2 fork node using the shell scripts `runL1TestNetwork.sh` and `runL2TestNetwork.sh` inside [script](../..) folder respectively.
```

ricott1 commented on src/L1/SwapAndBridge.sol:
```diff
@@ -0,0 +1,107 @@
+
+        // Check that this contract has no tokens left in its balance.
+        require(L1_TOKEN.balanceOf(address(this)) == 0, "Contract still has tokens.");
+    }
```
I think this check is fine, not sure how to improve it @gkoumout @matjazv

matjazv replied:
I think it's fine to mark it as `Won't fix`.

ricott1 committed:
Update script/swap_and_bridge/example/README.md

Co-authored-by: Matjaz Verbole <verbole@gmail.com>

ricott1 committed:
Update pr.yaml

ricott1 committed:
improving deployment script

ricott1 committed:
improving testing scripts

ricott1 committed:
improve test scripts and rename folder

ricott1 committed:
Merge branch 'development' into feature/swapAndBridge

matjazv commented on script/7_deployWdivETHContract.sh:
```diff
@@ -0,0 +1,17 @@
+echo "Done."
+
+echo "Deploying OptimismMintableERC20 wdivETH token smart contract..."
+cast send --rpc-url $L2_RPC_URL "0x4200000000000000000000000000000000000012" "createOptimismMintableERC20WithDecimals(address,string,string,uint8)" $L1_TOKEN_ADDR_DIVA "Wrapped Diva Ether Token" wdivETH 18 -i
```
Should `wdivETH` also be inside double quotes because it's a string like  "Wrapped Diva Ether Token"?

`-i`: why it needs to be in interactive mode? We have a deployer private key inside `.env` file.

ricott1 replied:
I think the double quotes are irrelevant, I will test adding them as well. It does not need to be interactive but I wanted to have the same flow as the USDT token, no preference from my side

matjazv replied:
I've overlooked `-i` when I was reviewing USDT token deployment script.
I was quite surprised what to do when we're deploying it and I observed script requires input from user.

Would be better to do like it's in other scripts where deployer private key is taken from `.env` file.

ricott1 replied:
Sounds good, should we also open a PR to change the USDT deployment as well?

matjazv replied:
As you prefer.

I think we won't use USDT deployment script for `Mainnet` deployment anymore, right?
But for the (potential) future use, we may change it to be more aligned with other deployment scripts.

ricott1 replied:
BTW, how do you handle it in other deployment scripts? I see that there is no private key specified there

matjazv replied:
Inside deployment script we import private key from `.env` file.
It's the same in all deployments scripts.
Example:
https://github.com/LiskHQ/lisk-contracts/blob/development/script/contracts/L2/L2Airdrop.s.sol#L20-#L21
and its usage:
https://github.com/LiskHQ/lisk-contracts/blob/development/script/contracts/L2/L2Airdrop.s.sol#L59-#L69

matjazv commented on script/swap_and_bridge/contracts/SwapAndBridge_e2e_test.s.sol:
Would be better to move this file inside `script/swap_and_bridge/test` folder since it's a test script?

ricott1 replied:
It is also a contract tho ;) but no preference from my side (@gkoumout ?)

matjazv replied:
[SwapAndBridge_e2e_test.s.sol](https://github.com/LiskHQ/lisk-contracts/pull/64/files/eab7b68c146f735bb69d97b2a5a70fd9b4463376#diff-5067cc8b5256beb8e0962b444ffa93eb0544f3c3fc20128134baa39ea89b41bc) won't be deployed anywhere, right? So we won't deploy it on `Mainnet`?

As far as I can see, smart contract inside this file is like other smart contracts inside `test` folder where we keep unit and other tests for other smart contracts?

ricott1 replied:
I moved it

matjazv commented on script/swap_and_bridge/contracts/SwapAndBridge_integration_test.s.sol:
Would be better to move this file inside `script/swap_and_bridge/test` folder since it's a test script?

matjazv commented on script/8_deploySwapAndBridge.sh:
```diff
@@ -0,0 +1,36 @@
+echo "Done."
+
+echo "Deploying and if enabled verifying SwapAndBridge smart contract for LIDO protocol..."
+forge script --rpc-url="$L1_RPC_URL" --broadcast --verify --verifier blockscout --verifier-url $L1_VERIFIER_URL -vvvv script/swap_and_bridge/contracts/SwapAndBridge.s.sol:SwapAndBridgeScript  $L1_LIDO_BRIDGE_ADDR $L1_TOKEN_ADDR_LIDO $L2_TOKEN_ADDR_LIDO --sig 'run(address,address,address)'
```
Probably I'm missing something, but where is `$L2_TOKEN_ADDR_LIDO` deployed?
We deploy `$L2_TOKEN_ADDR_DIVA` inside `7_deployWdivETHContract.sh` script.
I can't find where is Lido deployed on L2?

What about L1 networks deployment for Lido and Diva? Were they deployed by Lido and Diva teams themself?

ricott1 replied:
Regarding L1 deployments: they are deployed by Diva and Lido team, we have no control over that.

matjazv replied:
Ok, cool!
Should we then update `.env.mainnet` if we already know what's the address of L1 deployments on L1 Mainnet?

ricott1 replied:
These integrations have not been deployed on mainnet yet. Once that's done we can update the env file

matjazv replied:
Ok, nice!

What about this part for L2?
```
Probably I'm missing something, but where is $L2_TOKEN_ADDR_LIDO deployed?
We deploy $L2_TOKEN_ADDR_DIVA inside 7_deployWdivETHContract.sh script.
I can't find where is Lido deployed on L2?
```

ricott1 replied:
It's deployed in another feature, related but independent from this. The deployment is rather complex and we decided to track it independently for clarity and audit purposes. From the POV of this feature, it's just a parameter in the .env file

matjazv replied:
Thanks for clarification!

matjazv commented on script/swap_and_bridge/contracts/SwapAndBridge.s.sol:
```diff
@@ -0,0 +1,25 @@
+        assert(_l1Bridge != address(0));
+        assert(_l1Token != address(0));
+        assert(_l2Token != address(0));
+        // Deployer's private key. L1_SWAP_AND_BRIDGE_DEPLOYER_PRIVATE_KEY is set in .env file.
```
```suggestion
        // Deployer's private key. PRIVATE_KEY is set in .env file.
```

ricott1 replied:
fixed

matjazv commented on script/swap_and_bridge/test/README.md:
```diff
@@ -0,0 +1,11 @@
+
+## Integration Tests
+
+These tests run some checks on the itneraction between the `SwapAndBridge` contract and the Lido and Diva LSTs. Run them with `./integration_test.sh`.
```
```suggestion
These tests run some checks on the interaction between the `SwapAndBridge` contract and the Lido and Diva LSTs. Run them with `./integration_test.sh`.
```

ricott1 replied:
fixed

matjazv commented on script/swap_and_bridge/test/e2e_test.sh:
```diff
@@ -0,0 +1,53 @@
+    # The following command will revert if the token has already been deployed with the same parameters because the deterministic address would clash
+    DeploymentJSON=$(cast send --rpc-url "http://localhost:8546" "0x4200000000000000000000000000000000000012" "createOptimismMintableERC20WithDecimals(address,string,string,uint8)" $L1_TOKEN_ADDR_DIVA "Wrapped Diva Local Token" wdivETH 18 --private-key $PRIVATE_KEY)
+    L2_TOKEN_ADDR_DIVA="0x$(echo $DeploymentJSON  | awk 'BEGIN{FS="topics:*"}{print $2}' | awk -F, '{print $3}' | awk 'BEGIN{FS="\]*"}{print $1}' | cut -d "\"" -f 2 | tail -c 41)"
+    echo "L2_TOKEN_ADDR_DIVA deployed to" $L2_TOKEN_ADDR_DIVA
```
`$L2_TOKEN_ADDR_DIVA` may be also set inside `.env` file when this script will be executed for the 2nd time:
```
sed -i -e 's/L2_TOKEN_ADDR_DIVA=0x/L2_TOKEN_ADDR_DIVA=$L2_TOKEN_ADDR_DIVA' .env
```

ricott1 replied:
that's pretty sweet

matjazv commented on src/L1/SwapAndBridge.sol:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    /// @notice Swap ETH to wrapped LST and bridge it to the recipient address on the L2.
+    /// If the amount of l1 token obtained is less than minL1Tokens, the transaction will revert.
```
Second line needs to be aligned with `Swap ETH to wrapped...`.

matjazv commented on src/L1/SwapAndBridge.sol:
```diff
@@ -0,0 +1,108 @@
+///         It is designed to be used as a part of the Lisk L2 ecosystem.
+contract SwapAndBridge {
+    /// @notice Minimum amount of gas to be used for the deposit message on L2.
+    uint32 public constant MIN_DEPOSIT_GAS = 0;
```
Once again, why can be `MIN_DEPOSIT_GAS` set to zero?

ricott1 replied:
It will be overridden in the bridge to a minimum calculated from the message. Having a non-zero min gas can have the effect of having the user pay an extra, which I think we should not set here.

matjazv replied:
Should we add similar comment inside file for this variable?
I think this is not something reader will know in advance without having a deep knowledge about how bridging works.

ricott1 replied:
this is the relevant link btw:https://github.com/ethereum-optimism/optimism/blob/develop/packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol#L187

I will add a comment, but there is this detail that in principle here we only require the bridge to follow the `depositERC20To` interface, with no knowledge of the bridge internals. Then of course, the tandard bridge is an important example and we could add the comment anyway

matjazv commented on test/L1/SwapAndBridge.t.sol:
```diff
@@ -0,0 +1,169 @@
+        swapAndBridge.swapAndBridgeTo{ value: 1 }(address(0));
+    }
+
+    function test_ReceiveFallback(uint256 tokensPerETH, uint256 value) public {
```
```suggestion
    function testFuzz_ReceiveFallback(uint256 tokensPerETH, uint256 value) public {
```

All test functions below which are property based tests need to have `testFuzz` at the beginning.

ricott1 replied:
added

matjazv commented on test/L1/SwapAndBridge.t.sol:
```diff
@@ -0,0 +1,169 @@
+        MockStandardBridge bridge = new MockStandardBridge();
+        SwapAndBridge swapAndBridge = new SwapAndBridge(address(bridge), address(mockLST), address(0xbeef));
+
+        vm.startPrank(testAccount);
+        if (recipient != address(0)) {
+            swapAndBridge.swapAndBridgeTo{ value: value }(recipient);
+        }
+        vm.stopPrank();
```
We tested that `swapAndBridgeTo` function doesn't revert.
May we also check below if some other conditions are true like `testAccount balance (on L1, possibly also on L2?)`, `bridge balance`, etc.?
So that state after `swapAndBridgeTo` is executed is correct?

matjazv replied:
Same is true for other test functions below except the ones which test that `revert` occurs.

ricott1 replied:
testing the balance on the L2 is not possible unless we run an E2E script, as it requires also dispatching the message to the L2 node, but I think this is covered in the e2e tests. I added checks for the L1 state whenever possible

matjazv replied:
Ok, then it's fine how it is implemented.

gkoumout commented on test/L1/SwapAndBridge.t.sol:
```diff
@@ -0,0 +1,169 @@
+        MockStandardBridge bridge = new MockStandardBridge();
+        SwapAndBridge swapAndBridge = new SwapAndBridge(address(bridge), address(mockLST), address(0xbeef));
+
+        uint256 expectedAmount = value * tokensPerETH / 1e18;
```
Minor point: Perhaps add a comment that this is the expected amount of st tokens expected to get from the mock ERC20 contract.

ricott1 replied:
added

gkoumout commented on test/L1/SwapAndBridge.t.sol:
```diff
@@ -0,0 +1,169 @@
+    }
+
+    function test_InvalidMsgValue(uint256 tokensPerETH, uint256 minL1Tokens) public {
+        MockERC20LST mockLST = new MockERC20LST(tokensPerETH);
```
I guess here we don't even need to mock, we could just define 

` SwapAndBridge swapAndBridge = new SwapAndBridge(address(0xdead), address(0xbeef), address(0x1111));`

and then just call `swapAndBridge.swapAndBridgeTo{ value: 0 }(address(0xgreg))`. 

But not a big deal, so we can keep it as is

ricott1 replied:
I think consistency is better here, I find it more clear to what we are doing

gkoumout replied:
My comment was mostly to be consistent with `test_InvalidRecipient`. But either way is fine for me.

ricott1 replied:
true, but I think that would be less clear, so I would not change it

gkoumout commented on test/L1/SwapAndBridge.t.sol:
```diff
@@ -0,0 +1,169 @@
+        vm.expectRevert("No wrapped tokens minted.");
+        swapAndBridge.swapAndBridgeToWithMinimumAmount{ value: value }(testAccount, 1e18);
+        vm.stopPrank();
+    }
```
Tests for the following errors are missing:
  - "Failed to approve L1 token spender."
  - "Contract still has tokens."
  
  I guess we can add them by mocking fake functionalities of `approve` and `depositERC20To` functions so that they either i) approve higher amount than user balance ii) deposit smaller amount than msg.value.

ricott1 replied:
indeed I was not testing them as it would require introducing bugs in the Mock contracts basically. 

I added the fake contracts and relative tests.

gkoumout commented on test/L1/SwapAndBridge.t.sol:
```diff
@@ -0,0 +1,169 @@
+        vm.stopPrank();
+    }
+
+    function test_InsufficientTokensMinted(uint256 tokensPerETH, uint256 value) public {
```
Minor improvement point: I think this test could be merged with the previous one for making it shorter. Just have the three checks in the same test (e.g., lines 97-100):

>         vm.startPrank(testAccount);
>         swapAndBridge.swapAndBridgeToWithMinimumAmount{ value: value }(testAccount, expectedAmount - 1);
>         swapAndBridge.swapAndBridgeToWithMinimumAmount{ value: value }(testAccount, expectedAmount);
>                 vm.expectRevert("Insufficient L1 tokens minted.");
>         swapAndBridge.swapAndBridgeToWithMinimumAmount{ value: value }(testAccount, expectedAmount + 1);
>  
>         vm.stopPrank();

ricott1 replied:
this is true, but I would prefer to keep the tests as atomic and clear as possible

matjazv commented on script/swap_and_bridge/contracts/SwapAndBridge_e2e_test.s.sol:
```diff
@@ -0,0 +1,324 @@
+    function run(address _l2Token) public {
+        assert(_l2Token != address(0));
+        IWrappedETH l2LSTToken = IWrappedETH(payable(_l2Token));
+        string memory dataPath = string.concat(vm.projectRoot(), "/script/swap_and_bridge/example/message.data");
```
Should `message.data` file be deleted at the end of test like it's done in below link?
https://github.com/LiskHQ/lisk-contracts/blob/development/test/utils/Utils.t.sol#L31

ricott1 replied:
I do delete the file using `vm.removeFile(dataPath);` on line 310

matjazv replied:
You're right!
I didn't see it.

matjazv commented on script/swap_and_bridge/contracts/SwapAndBridge_integration_test.s.sol:
```diff
@@ -0,0 +1,88 @@
+        vm.startPrank(testAccount);
+
+        console2.log("Testing no minimum...");
+        swapAndBridge.swapAndBridgeToWithMinimumAmount{ value: 1 ether }(testAccount, 0);
```
Should we add some `require` after its execution to test if this function was really executed as expected?

ricott1 replied:
added

matjazv commented on script/swap_and_bridge/contracts/SwapAndBridge_integration_test.s.sol:
```diff
@@ -0,0 +1,88 @@
+        console2.log("Ok");
+
+        console2.log("Testing 'High enough limit'...");
+        swapAndBridge.swapAndBridgeToWithMinimumAmount{ value: 1 ether }(testAccount, 1e18);
```
Should we add some `require` after its execution to test if this function was really executed as expected?

ricott1 replied:
added

matjazv commented on script/swap_and_bridge/contracts/SwapAndBridge_integration_test.s.sol:
```diff
@@ -0,0 +1,88 @@
+
+    function runReceive() public {
+        console2.log("Testing converting ETH to LST token...");
+        vm.recordLogs();
```
Is `vm.recordLogs()` needed somewhere?

ricott1 replied:
no, this one can be removed

ricott1 committed:
Merge branch 'development' into feature/swapAndBridge

ricott1 committed:
Address review comments

matjazv commented on script/8_deploySwapAndBridge.sh:
```diff
@@ -0,0 +1,36 @@
+source .env
+echo "Done."
+
+echo "Creating $NETWORK directory inside deployment directory..."
+if [ -z "$NETWORK" ]
+then
+      echo "NETWORK variable inside .env file is not set. Please set NETWORK environment variable."
+      exit 1
+else
+      if [ -d "deployment/swap_and_bridge/$NETWORK" ]
+      then
+            echo "Directory deployment/swap_and_bridge/$NETWORK already exists."
+      else
+            mkdir -p deployment/swap_and_bridge/$NETWORK
+      fi
+fi
+echo "Done."
```
Is this actually needed?
Does any of below scripts read/write something inside `deployment` directory?
Otherwise it may be removed.

ricott1 replied:
removed

matjazv commented on script/swap_and_bridge/test/integration_test.sh:
```diff
@@ -0,0 +1,51 @@
+    exit 1
+fi
+
+if [ -z "$L2_TOKEN_ADDR_DIVA" ]
```
That this works properly `L2_TOKEN_ADDR_DIVA` need to be set inside `.env` files (example, mainnet) like this:
`L2_TOKEN_ADDR_DIVA=`
and not like it currently is:
`L2_TOKEN_ADDR_DIVA=0x`

If I put `L2_TOKEN_ADDR_DIVA=` inside my `.env` file I get following error:
```
L2_TOKEN_ADDR_DIVA variable inside .env file is not set.
Deploying 'Wrapped Diva Ether Token' on L2...
Error: 
server returned an error response: error code 3: execution reverted
```

ricott1 replied:
fixed

matjazv committed:
Fix contracts folder path inside deployment directory

ricott1 committed:
Fix typo in e2e test and move contracts to test folder

ricott1 committed:
remove unnecessary interaction with deployment folder

ricott1 committed:
Merge branch 'development' into feature/swapAndBridge

matjazv committed:
Fix contracts folder path inside deployment directory

ricott1 committed:
Merge branch 'development' of github.com:LiskHQ/lisk-contracts into development

ricott1 committed:
merge development

ricott1 committed:
add testing contracts

ricott1 committed:
merge

ricott1 committed:
merge from development

ricott1 committed:
fix sed rule

ricott1 committed:
remove store of L@_DIVA_TOKEN

matjazv commented on script/swap_and_bridge/test/README.md:
```diff
@@ -0,0 +1,13 @@
+Tests in this folder are run as script as they require to be run in a Sepolia forked environment, since they depend on the contracts owned and deployed by Lido and Diva teams.
+
+Before running them, start an L1 and L2 fork node using the shell scripts `runL1TestNetwork.sh` and `runL2TestNetwork.sh` inside [script](../..) folder respectively and make sure to run the tests from this folder.
+
+Notice that these tests may fail fo reasons unrelated to the implementation of this feature, but rather because of problems with the target integration (for instance if the target LST contract does not mint new LSTs).
```
Maybe mention somewhere in this README file that in case `L2_TOKEN_ADDR_DIVA` is not set inside `env` file, before each execution of `./e2e_test.sh` or `./integration_test.sh` both local networks need to be restarted otherwise deployment of Diva token will fail because it's already deployed.

ricott1 replied:
done

ricott1 committed:
Improve README file

matjazv commented on script/swap_and_bridge/test/README.md:
```diff
@@ -1,8 +1,10 @@
+# Tst Scripts for SwapAndBridge
```
`Tst` -> `Test`

matjazv reviewed (approved):
Just one small comment left.

Well done!

ricott1 committed:
add readme

ricott1 committed:
fix typo

ricott1 committed:
add readme

ricott1 committed:
change folder structure

matjazv commented on script/swap_and_bridge/README.md:
```diff
@@ -0,0 +1,15 @@
+
+Tests in this folder are run as script as they require to be run in a Sepolia forked environment, since they depend on the contracts owned and deployed by Lido and Diva teams.
+
+Before running them, start an L1 and L2 fork node using the shell scripts `runL1TestNetwork.sh` and `runL2TestNetwork.sh` inside [script](../..) folder respectively and make sure to run the tests from this folder.
```
```suggestion
Before running them, start an L1 and L2 fork node using the shell scripts `runL1TestNetwork.sh` and `runL2TestNetwork.sh` inside [script](..) folder respectively and make sure to run the tests from this folder.
```

ricott1 committed:
Update script/swap_and_bridge/README.md

Co-authored-by: Matjaz Verbole <verbole@gmail.com>

ricott1 committed:
Merge branch 'development' into feature/swapAndBridge

gkoumout commented on documentation/swap-and-bridge.md:
```diff
@@ -0,0 +1,54 @@
+
+## Motivation
+
+Liquid staking pools allow users to participate in ETH staking without managing their own validator node or owning the full 32 ETH deposit amount. ETH from several users is pooled together and used to register a validator which then shares the rewards with the users. A Liquid Staking Token is an ERC20 token representing one of this staking position. As such, it can be converted back to ETH, while ...
```
Minor suggestion for improvement:
```suggestion
Liquid staking pools allow users to participate in ETH staking without managing their own validator node or owning the full 32 ETH deposit amount. ETH from several users is pooled and used to register a validator which then shares the rewards with them. A Liquid Staking Token is an ERC20 token representing one of those staking positions. As such, it can be converted back to ETH, while at the same time being freely tradable as an independent token. Furthermore, this token provides the extra yield from the validators reward either by rebasing, i.e. changing the user balance to account for the extra tokens acquired, or, in their wrapped version, by changing the underlying conversion rate to ETH.
```

ricott1 replied:
edited in this direction

gkoumout commented on documentation/swap-and-bridge.md:
```diff
@@ -0,0 +1,54 @@
+
+The swap-and-bridge feature prefers to interact with the wrapped version of a LST, as this tends to be better supported on the UX of several DeFi protocols (see the discussion for Lido's [wstETH](https://docs.lido.fi/contracts/wsteth/#why-use-wsteth)) and since the Optimism standard bridge [does not support rebasing tokens correctly](https://docs.optimism.io/builders/app-developers/bridging/stand ...
+
+A similar result for the end-user (swapping L1 ETH for L2 wrapped LST with a single interaction) can be achieved by using one of the several bridges available. However, this flow likely offers better security guarantees by relying on the L1<-->L2 bridging protocol and also stakes the ETH in the underlying staking pool protocol. Because of this, it should be considered an alternative option to sti ...
```
I am not 100% sure what do you mean by the first sentence of using bridges to achieve the same result.

ricott1 replied:
Liquidity bridges support swapping and bridging with a single interaction, but the underlying mechanism is different

gkoumout commented on documentation/swap-and-bridge.md:
```diff
@@ -0,0 +1,54 @@
+A similar result for the end-user (swapping L1 ETH for L2 wrapped LST with a single interaction) can be achieved by using one of the several bridges available. However, this flow likely offers better security guarantees by relying on the L1<-->L2 bridging protocol and also stakes the ETH in the underlying staking pool protocol. Because of this, it should be considered an alternative option to sti ...
+
+## Rationale
+
```
Perhaps start by saying "For each wrapped LST supported, a different `SwapAndBridge` contract needs to be deployed."

ricott1 replied:
edited in this direction

gkoumout commented on documentation/swap-and-bridge.md:
```diff
@@ -0,0 +1,54 @@
+
+This repository does not include any UI-related code.
+
+## Incident response plan
```
Do we really need this paragraph?

ricott1 replied:
I added it explicitly to address a potential audit. Maybe we could rename it to something different?

gkoumout replied:
Ok, no problem from my side, just asking since for other contracts we don't have a similar section.

ricott1 committed:
update documentation

ricott1 committed:
Merge branch 'feature/swapAndBridge' of github.com:LiskHQ/lisk-contracts into feature/swapAndBridge

ricott1 committed:
Fix slither warning
